Mathlib/NumberTheory/EulerProduct/Basic.lean

```diff
@@ -0,0 +1,166 @@
+    (h : ∀ {p : ℕ}, p.Prime → ‖f p‖ < 1) (N : ℕ) :
+    Summable (fun m : N.smoothNumbers ↦ ‖f m‖) ∧
+      HasSum (fun m : N.smoothNumbers ↦ f m) (∏ p in N.primesBelow, (1 - f p)⁻¹) := by
+  have hmul {m n} (_ : Nat.Coprime m n) := f.map_mul m n
+  convert summable_and_hasSum_smoothNumbers_prod_primesBelow_tsum f.map_one hmul ?_ N with M hM <;>
+    simp_rw [map_pow]
+  · exact (tsum_geometric_of_norm_lt_1 <| h <| prime_of_mem_primesBelow hM).symm
+  · intro p hp
+    simp_rw [norm_pow]
+    exact summable_geometric_iff_norm_lt_1.mpr <| (norm_norm (f p)).symm ▸ h hp
+
+-- We now assume in addition that `f` is norm-summable.
+variable (hsum : Summable (‖f ·‖))
+
+/-- A version of `EulerProduct.summable_and_hasSum_smoothNumbers_prod_primesBelow_tsum`
+in terms of the value of the series. -/
+lemma prod_primesBelow_tsum_eq_tsum_smoothNumbers (N : ℕ) :
+    ∏ p in N.primesBelow, ∑' (n : ℕ), f (p ^ n) = ∑' m : N.smoothNumbers, f m :=
+  (summable_and_hasSum_smoothNumbers_prod_primesBelow_tsum hf₁ hmul
+    (fun hp ↦ hsum.comp_injective <| Nat.pow_right_injective hp.one_lt) _).2.tsum_eq.symm
+
+/-- A version of `EulerProduct.summable_and_hasSum_smoothNumbers_prod_primesBelow_geometric`
+in terms of the value of the series. -/
+lemma prod_primesBelow_geometric_eq_tsum_smoothNumbers {f : ℕ →* F} (hsum : Summable f) (N : ℕ) :
+    ∏ p in N.primesBelow, (1 - f p)⁻¹ = ∑' m : N.smoothNumbers, f m := by
+  refine (summable_and_hasSum_smoothNumbers_prod_primesBelow_geometric ?_ N).2.tsum_eq.symm
+  exact fun {_} hp ↦ hsum.norm_lt_one hp.one_lt
+
+/-- The following statement says that summing over `N`-smooth numbers
+for large enough `N` gets us arbitrarily close to the sum over all natural numbers
+(assuming `f` is norm-summable and `f 0 = 0`; the latter since `0` is not smooth). -/
+lemma norm_tsum_smoothNumbers_sub_tsum_lt (hsum : Summable f) (hf₀ : f 0 = 0)
+    {ε : ℝ} (εpos : 0 < ε) :
+    ∃ N₀ : ℕ, ∀ N ≥ N₀, ‖(∑' m : ℕ, f m) - (∑' m : N.smoothNumbers, f m)‖ < ε := by
+  obtain ⟨N₀, hN₀⟩ :=
+    summable_iff_nat_tsum_vanishing.mp hsum (Metric.ball 0 ε) <| Metric.ball_mem_nhds 0 εpos
+  simp_rw [mem_ball_zero_iff] at hN₀
+  refine ⟨N₀, fun N hN₁ ↦ ?_⟩
+  convert hN₀ _ <| N.smoothNumbers_compl.trans fun _ ↦ hN₁.le.trans
+  simp_rw [← tsum_subtype_add_tsum_subtype_compl hsum N.smoothNumbers,
+    add_sub_cancel', tsum_eq_tsum_diff_singleton (N.smoothNumbers)ᶜ hf₀]
+
+open Filter in
+/-- The *Euler Product* for multiplicative (on coprime arguments) functions.
+If `f : ℕ → F`, where `F` is a complete normed field, `f 0 = 0`,
+`f 1 = 1`, `f` is multiplicative on coprime arguments,
+and `‖f ·‖` is summable, then `∏' p : {p : ℕ | p.Prime}, ∑' e, f (p ^ e) = ∑' n, f n`.
+Since there are no infinite products yet in Mathlib, we state it in the form of
+convergence of finite partial products. -/
+-- TODO: Change to use `∏'` once infinite products are in Mathlib
+theorem euler_product (hf₀ : f 0 = 0) :
```

Comment:
Since "Euler product" is atomic, namingwise, I suggest to call this `eulerProduct`.
```suggestion
theorem eulerProduct (hf₀ : f 0 = 0) :
```
If you agree, please make the corresponding adaptation below.

Reply:
Will do. (Breathing a sigh of relief that you didn't propose `tendsto_prod_primesBelow_tsum_nat_comp_pow_atTop_tsum_nat_of_multiplicative_of_summable_norm` or so :smile: )

Reply:
OK.